presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java

```diff
@@ -156,4 +157,14 @@ public Expression rewriteExtract(Extract node, Void context, ExpressionTreeRewri
             throw new UnsupportedOperationException("not yet implemented: " + node.getField());
         }
     }
+
+    private static boolean isLiteral(Expression expression)
+    {
+        // Current IR has no way to represent typed constants, so it encodes them as Cast(Literal)
+        if (expression instanceof Cast && ((Cast) expression).getExpression() instanceof Literal) {
+            return true;
```

Comment:
This is not sufficient for:

https://github.com/prestosql/presto/blob/1116011e9f9e0a5b54740b349c3476412d6afdba/presto-main/src/main/java/io/prestosql/sql/planner/LiteralEncoder.java#L134-L138
(DOUBLE, REAL ±infinity)

https://github.com/prestosql/presto/blob/1116011e9f9e0a5b54740b349c3476412d6afdba/presto-main/src/main/java/io/prestosql/sql/planner/LiteralEncoder.java#L204-L212
(anything Block-based, varbinary, json or anything Slice-based other than varchar, char and decimal)

Reply:
I'm aware. I wanted to do the simplest possible thing that 1) doesn't require ExpressionInterpreter.optimize, 2) doesn't try to duplicate all the logic in LiteralEncoder, 3) covers a sufficient portion of the use cases that occur in practice

Reply:
Sure. Please expand your comment. `// Current IR has no way to represent typed constants, so it encodes them as Cast(Literal` sounds like these other cases are being overlooked, rather than consciously ignored.